python-checks/src/main/java/org/sonar/python/checks/UndefinedNameAllPropertyCheck.java

```diff
@@ -0,0 +1,162 @@
+import org.sonar.plugins.python.api.tree.Name;
+import org.sonar.plugins.python.api.tree.StringElement;
+import org.sonar.plugins.python.api.tree.StringLiteral;
+import org.sonar.plugins.python.api.tree.Tree;
+import org.sonar.plugins.python.api.tree.Tuple;
+import org.sonar.python.semantic.BuiltinSymbols;
+import org.sonar.python.tree.TreeUtils;
+
+@Rule(key = "S5807")
+public class UndefinedNameAllPropertyCheck extends PythonSubscriptionCheck {
+
+  @Override
+  public void initialize(Context context) {
+    context.registerSyntaxNodeConsumer(Tree.Kind.ASSIGNMENT_STMT, ctx -> {
+      AssignmentStatement assignmentStatement = (AssignmentStatement) ctx.syntaxNode();
+      if (TreeUtils.firstAncestorOfKind(assignmentStatement, Tree.Kind.CLASSDEF, Tree.Kind.FUNCDEF) != null) {
+        // We only consider __all__ assignment at module level
+        return;
+      }
+      ExpressionList expressionList = assignmentStatement.lhsExpressions().get(0);
+      if (expressionList.expressions().size() > 1) {
+        return;
+      }
+      Expression lhs = expressionList.expressions().get(0);
+      if (lhs.is(Tree.Kind.NAME) && ((Name) lhs).name().equals("__all__")) {
+        checkAllProperty(ctx, assignmentStatement);
+      }
+    });
+  }
+
+  private static void checkAllProperty(SubscriptionContext ctx, AssignmentStatement assignmentStatement) {
+    Expression assignedValue = assignmentStatement.assignedValue();
+    if (!assignedValue.is(Tree.Kind.LIST_LITERAL, Tree.Kind.TUPLE)) {
+      return;
+    }
+    List<Tree> stringExpressions = getStringExpressions(assignedValue);
+    FileInput fileInput = (FileInput) TreeUtils.firstAncestorOfKind(assignedValue, Tree.Kind.FILE_INPUT);
+    if (fileInput == null) {
+      // Should never happen
+      return;
+    }
+    UnresolvedWildcardImportVisitor unresolvedWildcardImportVisitor = new UnresolvedWildcardImportVisitor();
+    fileInput.accept(unresolvedWildcardImportVisitor);
+    if (unresolvedWildcardImportVisitor.hasUnresolvedWildcardImport) {
+      return;
+    }
+    Map<String, Symbol> symbolsByName = fileInput.globalVariables().stream().collect(Collectors.toMap(Symbol::name, Function.identity()));
+    for (Tree stringExpression : stringExpressions) {
+      StringLiteral stringLiteral = retrieveStringLiteral(stringExpression);
+      if (isUnknownSymbol(ctx.pythonFile(), symbolsByName, stringLiteral)) {
+        PreciseIssue issue = ctx.addIssue(stringExpression, String.format("Change or remove this string; \"%s\" is not defined.", stringLiteral.trimmedQuotesValue()));
+        if (stringExpression != stringLiteral) {
+          issue.secondary(stringLiteral, null);
+        }
+      }
+    }
+  }
+
+  private static boolean isUnknownSymbol(PythonFile pythonFile, Map<String, Symbol> symbolsByName, StringLiteral stringLiteral) {
+    String name = stringLiteral.trimmedQuotesValue();
+    return stringLiteral.stringElements().stream().noneMatch(StringElement::isInterpolated) &&
```

Comment:
_minor_ I would move the `&&` operator to the next line, as it is done for the other lines.